src/Domains/Registrar/OpenSRS.php

```diff
@@ -0,0 +1,647 @@
+        }
+
+        $result = $this->send($message);
+
+        return $result;
+    }
+
+    public function purchase(string $domain, array $contacts, array $nameservers = []): array
+    {
+        $nameservers =
+          empty($nameservers)
+          ? $this->defaultNameservers
+          : $nameservers;
+
+        $contacts = $this->sanitizeContacts($contacts);
+
+        $regType = 'new';
+
+        $result = $this->register($domain, $regType, $this->user, $contacts, $nameservers);
+
+        $result = $this->response($result);
+
+        return $result;
+    }
+
+    public function transfer(string $domain, array $contacts, array $nameservers = []): array
+    {
+        $nameservers =
+          empty($nameservers)
+          ? $this->defaultNameservers
+          : $nameservers;
+
+        $contacts = $this->sanitizeContacts($contacts);
+
+        $regType = 'transfer';
+
+        $result = $this->register($domain, $regType, $this->user, $contacts, $nameservers);
+        $result = $this->response($result);
+
+        return $result;
+    }
+
+    public function cancelPurchase(): bool
+    {
+        $timestamp = date('Y-m-d\TH:i:s.000');
+        $timestamp = strtotime($timestamp);
+
+        $message = [
+            'object' => 'ORDER',
+            'action' => 'cancel_pending_orders',
+            'attributes' => [
+                'to_date' => $timestamp,
+                'status' => [
+                    'declined',
+                    'pending',
+                ],
+            ],
+        ];
+
+        $result = $this->send($message);
+        $result = $this->sanitizeResponse($result);
+
+        $elements = $result->xpath('//body/data_block/dt_assoc/item[@key="is_success"]');
+        $successful = "{$elements[0]}" === '1' ? true : false;
+
+        return $successful;
+    }
+
+    public function suggest(array|string $query, array $tlds = [], $minLength = 1, $maxLength = 100): array
+    {
+        $query = is_array($query) ? $query : [$query];
+
+        $message = [
+            'object' => 'DOMAIN',
+            'action' => 'name_suggest',
+            'attributes' => [
+                'services' => [
+                    'suggestion',
+                ],
+                'searchstring' => implode(' ', $query),
+                'tlds' => $tlds,
+            ],
+        ];
+
+        $xpath = implode('/', [
+            '//body',
+            'data_block',
+            'dt_assoc',
+            'item[@key="attributes"]',
+            'dt_assoc',
+            'item[@key="suggestion"]',
+            'dt_assoc',
+            'item[@key="items"]',
+            'dt_array',
+            'item',
+        ]);
+
+        $result = $this->send($message);
+        $result = $this->sanitizeResponse($result);
+        $elements = $result->xpath($xpath);
+
+        $items = [];
+
+        foreach ($elements as $element) {
+            $item = $element->xpath('dt_assoc/item');
+            $domain = "{$item[0]}";
```

Comment:
I think casting like this would be more readable over all and also tells the story better why we do this
```suggestion
            $domain = (string) $item[0];
```